### PR TITLE
More data

### PR DESCRIPTION
**Pull request by dklong-csu**
The check is not going to pass. Locally, I fail only on likelihoodReasonable. The way I validated the results originally was by solving the ODE in MatLab with Derek's code, which is solved at a much higher accuracy. Then I just gave the resulting probability mass function the thumb test. These "new" values might pass the thumb test as well, but I don't think they should have changed due to my updates. I'm pretty sure the error is associated with the ODE solve, but I can't think of a good way to do a back-of-the-envelope check without doing each update step in a spreadsheet (but that'll have large roundoff errors). I'm still looking at this, but if you notice something in the meantime, that would be appreciated.

### CODE REVIEW

**Comment by dklong-csu** [comment]
I'm very confused. I tracked the issue down to the ODE solver evaluating to an unexpected value. This begins at the first time step. I rewrote the rhs from scratch and the result is unchanged. I'm comparing to calculations in a spreadsheet computed with formulas and to calculations manually typed in Matlab -- the spreadsheet and Matlab have the same results and I am treating those as "correct".

File: main.cpp [hunk 1]
```diff
@@ -31,6 +32,7 @@ using VectorType = std::valarray<double>;
 double log_likelihood (const SampleType &prm)
 {
   // S3 data
+  /*
```
Comment (bangerth) [inline main.cpp:35]: If you don't need something any more, just remove it. You can always get it back through the git history.

**Review by bangerth — COMMENTED** [review]
I've just taken a very quick look at this, but you're putting too many things into one pull request to see where things go wrong. Create one pull request that replaces the existing data arrays by the ones you get from `data.h`. Run that by the testers, and if it succeeds with the tests, merge that. Then move on to the next piece. Let each pull request be exactly one thing you add, rather than replacing data sources and a whole lot of other things. This way, you can check each thing individually. For example, in a second you can extend the interface of the functions that compute likelihoods to also take multiple time points, but just call them with the one time point you had previously and make sure that you get the expected (old) output.

It's possible that you're doing this step-by-step approach in the commits you have collected here. In that case, go back to `master`, create a new branch for your first pull request, and use `git cherry-pick` to isolate the commits you'd like to consider for that first pull request. Then upload, and let's look at this in isolation.

File: main.cpp [hunk 2]
```diff
@@ -42,9 +44,17 @@ double log_likelihood (const SampleType &prm)
                                           3.19, 1.73, 1.94, 2.14, 2.91, 2.85,
                                           2.8, 2.37, 2.42, 2.68, 2.01, 1.9,
                                           2.14};
+*/
+  const Data::PomData pom_data;
+  const std::valarray<double> dataDiam = pom_data.tem_diam_time2;
 
+  std::vector<std::valarray<double>> fitData(1);
```
Comment (bangerth) [inline main.cpp:51]:
```suggestion
  std::vector<std::valarray<double>> fitData = { std::pow(dataDiam/0.3000805, 3) };
```
This way you initialize right away and can, if you want, make the object `const`.

File: main.cpp [hunk 3]
```diff
@@ -42,9 +44,17 @@ double log_likelihood (const SampleType &prm)
...
+  fitData[0] = std::pow(dataDiam/0.3000805, 3);
 
-  const std::valarray<double> dataAtoms = std::pow(dataDiam/0.3000805, 3);
+  std::vector<double> fitTime(1);
+  fitTime[0] = pom_data.tem_time2;
```
Comment (bangerth) [inline main.cpp:55]: same approach here

File: models.cpp [hunk 4]
```diff
@@ -554,37 +555,50 @@ std::valarray<double> Models::ThreeStepAlternative::right_hand_side(const std::v
...
-  f[0] -= three_step_alt_parameters.k1 * x[0] * x[1] * x[1];
-  // dissasociated precursor -- loss from nucleation
+
+  // precursor
+    // loss from nucleation
+    // loss from particle growth
+    // loss from forward dissociative step
+    // gain from backward dissociative step
+  f[0] = -three_step_alt_parameters.k1 * x[0] * x[1] * x[1];
+  for (unsigned int i = 4; i < three_step_alt_parameters.n_variables; i++)
```
Comment (bangerth) [inline models.cpp:565]: This also looks like a change that you could isolate into a separate pull request and have it tested separately.

File: models.cpp [hunk 5]
```diff
@@ -596,41 +610,50 @@ std::valarray<double> Models::ThreeStepAlternative::right_hand_side(const std::v
 
 // class to hold integration hyperparameters
 Models::explEulerParameters::explEulerParameters(const double startTimeValue,
-                      const double endTimeValue,
+                      const std::vector<double> outputTimesValue,
```
Comment (bangerth) [inline models.cpp:613]:
```suggestion
                      const std::vector<double> &outputTimesValue,
```

File: models.cpp [hunk 6]
```diff
@@ -596,41 +610,50 @@ std::valarray<double> Models::ThreeStepAlternative::right_hand_side(const std::v
...
-    if (time + time_step > solverParameters.endTime)
+    while (time < evalTimes[i])
     {
-      time_step = solverParameters.endTime - time;
-      time = solverParameters.endTime;
+      //double endTime = evalTimes[i];
+      // advance to next time
+      if (time + dt > evalTimes[i])
+      {
+        dt = evalTimes[i] - time;
```
Comment (bangerth) [inline models.cpp:644]: You modify the time step here in order to get to the preset evaluation times. But do you ever recover it to its previous value for the following time steps?

File: models.cpp [hunk 7]
```diff
@@ -596,41 +610,50 @@ std::valarray<double> Models::ThreeStepAlternative::right_hand_side(const std::v
...
 // ODE solver
-std::valarray<double> Models::integrate_ode_explicit_euler(const Models::explEulerParameters solverParameters,
-                                                   const Models::ModelsBase& model,
-                                                   const Models::ParametersBase& modelParameters)
+std::vector<std::valarray<double>> Models::integrate_ode_explicit_euler(const Models::explEulerParameters solverParameters,
+                                                           const Models::ModelsBase& model,
+                                                           const Models::ParametersBase& modelParameters,
+                                                           const double time_step)
 {
+  const std::vector<double> evalTimes = solverParameters.outputTimes;
```
Comment (bangerth) [inline models.cpp:629]: You are, in essence, creating a duplicate of the variable you already have. Why not work with the original variable instead?

File: models.cpp [hunk 8]
```diff
@@ -596,41 +610,50 @@ std::valarray<double> Models::ThreeStepAlternative::right_hand_side(const std::v
...
+
+    std::vector<std::valarray<double>> solution(evalTimes.size());
   std::valarray<double> x = solverParameters.initialCondition;
-  double time_step = 1e-5;
 
   double time = solverParameters.startTime;
-  while (time < solverParameters.endTime)
+  double dt = time_step;
+  for (unsigned int i = 0; i < evalTimes.size(); i++)
```
Comment (bangerth) [inline models.cpp:636]: Maybe better to use a more descriptive variable name, in view of the discussion this morning? ;-)

File: models.h [hunk 9]
```diff
@@ -354,20 +355,22 @@ namespace Models
...
   public:
-    double startTime, endTime;
+    double startTime;
+    std::vector<double> outputTimes;
     std::valarray<double> initialCondition;
 
     // constructor
     explEulerParameters(const double startTimeValue,
-                        const double endTimeValue,
+                        const std::vector<double> outputTimes,
```
Comment (bangerth) [inline models.h:364]:
```suggestion
                        const std::vector<double> &outputTimes,
```

File: statistics.cpp [hunk 10]
```diff
@@ -10,14 +11,15 @@
 // This function solves the ODE based on odeModel, odeParameters, and solverParameters
 // Then the ODE solution and data are converted to histograms based on histogramParameters.
 // Finally the log likelihood is computed using the multinomial distribution and the two histograms.
-double Statistics::logLikelihood(const std::valarray<double>& particleSizeData,
-                                 const Models::ModelsBase& odeModel,
-                                 const Models::ParametersBase& odeParameters,
-                                 const Models::explEulerParameters& solverParameters,
-                                 const Histograms::Parameters& histogramParameters)
+std::tuple<double, std::vector<std::valarray<double>>, std::vector<std::valarray<double>>> Statistics::logLikelihoodInternal(const std::vector<std::valarray<double>>& particleSizeData,
```
Comment (bangerth) [inline statistics.cpp:14]: you might want to document what this awkward return object actually is

File: statistics.cpp [hunk 11]
```diff
@@ -27,42 +29,86 @@ double Statistics::logLikelihood(const std::valarray<double>& particleSizeData,
...
-  const double normalizeConst = concentrations.sum();
-  const std::valarray<double> particleSizePMF = concentrations/normalizeConst;
+  // start with log likelihood at 0
+  double logLikelihoodVal = 0.0;
 
-  // create histogram from PMF to aggregate bin probability
-  Histograms::Histogram odeHistogram(histogramParameters);
+  // instantiate vectors for future output
+  std::vector<std::valarray<double>> pmfs;
+  std::vector<std::valarray<double>> data_counts;
```
Comment (bangerth) [inline statistics.cpp:43]: Do you actually carry their values from one iteration of the loop below to the next iteration? If not, then move these variables' declarations into the loop, just above where you first need them -- remember the section about the lifetime of objects?

File: statistics.cpp (hunk 10 above)
Comment (bangerth) [reply statistics.cpp:14]: This change to the loglikelihood function is also conceptually separate from the remaining changes. You could put this into a separate PR as well and see what the tests have to say about it.

File: statistics.h [hunk 12]
```diff
@@ -14,11 +15,23 @@ namespace Statistics
...
                        const Models::explEulerParameters& solverParameters,
-                       const Histograms::Parameters& histogramParameters);
+                       const Histograms::Parameters& histogramParameters,
+                       const double& timeStep);
+
+
+
+  // Internal version of logLikelihood which returns intermediate steps for checking.
+  // Intermediate steps included: binned data counts, binned probability mass function
+  std::tuple<double, std::vector<std::valarray<double>>, std::vector<std::valarray<double>>> logLikelihoodInternal(const std::vector<std::valarray<double>>& particleSizeData,
```
Comment (bangerth) [inline statistics.h:29]: this is actually where the documentation of what's returned would go.

**Comment by dklong-csu** [comment]
This PR is deprecated now due to PR's 41-44, 46, and (soon to be) 47.

**Pull request closed without merging**